spring-cloud-data-admin/src/main/java/org/springframework/cloud/data/admin/controller/ModuleController.java

```diff
@@ -0,0 +1,158 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.data.admin.controller;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.cloud.data.core.ModuleCoordinates;
+import org.springframework.cloud.data.core.ModuleType;
+import org.springframework.cloud.data.module.registry.ModuleRegistration;
+import org.springframework.cloud.data.module.registry.ModuleRegistry;
+import org.springframework.cloud.data.rest.resource.DetailedModuleRegistrationResource;
+import org.springframework.cloud.data.rest.resource.ModuleRegistrationResource;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.web.PagedResourcesAssembler;
+import org.springframework.hateoas.ExposesResourceFor;
+import org.springframework.hateoas.PagedResources;
+import org.springframework.hateoas.mvc.ResourceAssemblerSupport;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * Handles all Module related interactions.
+ *
+ * @author Glenn Renfro
+ * @author Mark Fisher
+ * @author Gunnar Hillert
+ * @author Eric Bottard
+ * @author Gary Russell
+ * @author Patrick Peralta
+ */
+@RestController
+@RequestMapping("/modules")
+@ExposesResourceFor(ModuleRegistrationResource.class)
+public class ModuleController {
+
+	private final Assembler moduleAssembler = new Assembler();
+
+	private final ModuleRegistry registry;
+
+	@Autowired
+	public ModuleController(ModuleRegistry registry) {
+		this.registry = registry;
+	}
+
+	/**
+	 * List module registrations.
+	 */
+	@RequestMapping(method = RequestMethod.GET)
+	@ResponseStatus(HttpStatus.OK)
+	public PagedResources<? extends ModuleRegistrationResource> list(Pageable pageable,
+			PagedResourcesAssembler<ModuleRegistration> assembler,
+			@RequestParam(value = "type", required = false) ModuleType type,
+			@RequestParam(value = "detailed", defaultValue = "false") boolean detailed) {
+
+		List<ModuleRegistration> list = new ArrayList<>(registry.findAll());
+		if (type != null) {
+			for (Iterator<ModuleRegistration> iterator = list.iterator(); iterator.hasNext();) {
+				if (iterator.next().getType() != type) {
+					iterator.remove();
+				}
+			}
+		}
+
+		// modules have a "natural" sort order of type followed by name,
+		// use that instead of what "pageable" indicates as a sort order
+		Collections.sort(list);
```

Comment:
That doesn't make sense to me. If we provide a sort order, then it should be honored. If we don't want to support that, then let's drop the pageable.